File: src/main/java/com/medicalsystem/model/IdComparableEntity.java
package com.medicalsystem.model;

import javax.persistence.MappedSuperclass;
import java.util.Objects;

@MappedSuperclass
public abstract class IdComparableEntity {

    public abstract long getId();

    @Override
    public int hashCode() {
        return Objects.hash(getId(), getClass());
    }

    @Override
    public boolean equals(Object obj) {
        if (this == obj) {
            return true;
        }

        if (obj instanceof IdComparableEntity) {
            IdComparableEntity entity = (IdComparableEntity) obj;
            return Objects.equals(getClass(), entity.getClass()) && Objects.equals(getId(), entity.getId());
        }

        return false;
    }
}

File: src/main/kotlin/com/medicalsystem/domain/id/IdComparableEntity.kt
package com.medicalsystem.domain.id

import com.fasterxml.jackson.annotation.JsonIgnore
import java.util.*

/**
 * An abstract class providing an ID and methods for comparing entities using this ID.
 * Every managed entity should extend this class.
 *
 * @param ID name of the ID formField
 */
abstract class IdComparableEntity<ID>(@JsonIgnore var defaultId: ID? = null) {
    abstract var id: ID

    override fun equals(other: Any?): Boolean =
            if (other is IdComparableEntity<*>) {
                id != defaultId && Objects.equals(id, other.id)
            } else {
                false
            }

    override fun hashCode(): Int = 31
}